ansible_base/lib/dynamic_config/dynamic_settings.py

```diff
@@ -25,229 +29,41 @@
-    # entries mapping that permission to the assignment's organization
-    ANSIBLE_BASE_CACHE_PARENT_PERMISSIONS = False
-
-    # API clients can assign users and teams roles for shared resources
-    ALLOW_LOCAL_RESOURCE_MANAGEMENT = True
-
-    try:
-        MANAGE_ORGANIZATION_AUTH
-    except NameError:
-        MANAGE_ORGANIZATION_AUTH = True
-
-    try:
-        ORG_ADMINS_CAN_SEE_ALL_USERS
-    except NameError:
-        ORG_ADMINS_CAN_SEE_ALL_USERS = True
-
-
-if 'ansible_base.oauth2_provider' in INSTALLED_APPS:  # noqa: F821
-    if 'oauth2_provider' not in INSTALLED_APPS:  # noqa: F821
-        INSTALLED_APPS.append('oauth2_provider')  # noqa: F821
-
-    try:
-        OAUTH2_PROVIDER  # noqa: F821
-    except NameError:
-        OAUTH2_PROVIDER = {}
-
-    if 'ACCESS_TOKEN_EXPIRE_SECONDS' not in OAUTH2_PROVIDER:
-        OAUTH2_PROVIDER['ACCESS_TOKEN_EXPIRE_SECONDS'] = 31536000000
-    if 'AUTHORIZATION_CODE_EXPIRE_SECONDS' not in OAUTH2_PROVIDER:
-        OAUTH2_PROVIDER['AUTHORIZATION_CODE_EXPIRE_SECONDS'] = 600
-    if 'REFRESH_TOKEN_EXPIRE_SECONDS' not in OAUTH2_PROVIDER:
-        OAUTH2_PROVIDER['REFRESH_TOKEN_EXPIRE_SECONDS'] = 2628000
-    if 'PKCE_REQUIRED' not in OAUTH2_PROVIDER:
-        # For compat with awx, we don't require PKCE, but the new version
-        # of DOT that we are using requires it by default.
-        OAUTH2_PROVIDER['PKCE_REQUIRED'] = False
-
-    OAUTH2_PROVIDER['APPLICATION_MODEL'] = 'dab_oauth2_provider.OAuth2Application'
-    OAUTH2_PROVIDER['ACCESS_TOKEN_MODEL'] = 'dab_oauth2_provider.OAuth2AccessToken'
-
-    oauth2_authentication_class = 'ansible_base.oauth2_provider.authentication.LoggedOAuth2Authentication'
-    if 'DEFAULT_AUTHENTICATION_CLASSES' not in REST_FRAMEWORK:  # noqa: F821
-        REST_FRAMEWORK['DEFAULT_AUTHENTICATION_CLASSES'] = []  # noqa: F821
-    if oauth2_authentication_class not in REST_FRAMEWORK['DEFAULT_AUTHENTICATION_CLASSES']:  # noqa: F821
-        REST_FRAMEWORK['DEFAULT_AUTHENTICATION_CLASSES'].insert(0, oauth2_authentication_class)  # noqa: F821
-
-    # These have to be defined for the migration to function
-    OAUTH2_PROVIDER_APPLICATION_MODEL = 'dab_oauth2_provider.OAuth2Application'
-    OAUTH2_PROVIDER_ACCESS_TOKEN_MODEL = 'dab_oauth2_provider.OAuth2AccessToken'
-    OAUTH2_PROVIDER_REFRESH_TOKEN_MODEL = "dab_oauth2_provider.OAuth2RefreshToken"
-    OAUTH2_PROVIDER_ID_TOKEN_MODEL = "dab_oauth2_provider.OAuth2IDToken"
-
-    ALLOW_OAUTH2_FOR_EXTERNAL_USERS = False
+    OAUTH2_PROVIDER = {}
+
+
+for key, value in get_dab_settings(
+    installed_apps=INSTALLED_APPS,
+    rest_framework=REST_FRAMEWORK,
+    spectacular_settings=SPECTACULAR_SETTINGS,
+    authentication_backends=local_authentication_backends,
+    middleware=MIDDLEWARE,
+    oauth2_provider=OAUTH2_PROVIDER,
+).items():
+    if key in ANSIBLE_BASE_OVERRIDABLE_SETTINGS:
+        ANSIBLE_BASE_OVERRIDDEN_SETTINGS.append(key)
+        locals()[key] = value
```

Comment:
AFAIK, mutating `locals()` will not have side effects on local variables, is that the intention?

I have used `setattr(sys.modules[__name__], name, value)` to mutate local variables, but maybe you really want to avoid the global side effect it has.

Reply:
Did you mean that it _can_ have side effects on local variables? I'm aware that those hazards exist, I'm just not sure if they matter in this context, where we won't be referencing them later. We only need to set variables for another module to include and pick them up.

Reply:
nevermind, I just verified that `locals()[key] = value` has the same effect as `sys.modules[__name__].key = value`